src/core/lib/event_engine/windows/grpc_polled_fd_windows.cc

```diff
@@ -0,0 +1,821 @@
+                      EventEngine* event_engine)
+      : name_(absl::StrFormat("c-ares socket: %" PRIdPTR,
+                              winsocket->raw_socket())),
+        address_family_(address_family),
+        socket_type_(socket_type),
+        mu_(mu),
+        winsocket_(std::move(winsocket)),
+        read_buf_(grpc_empty_slice()),
+        write_buf_(grpc_empty_slice()),
+        outer_read_closure_([this]() { OnIocpReadable(); }),
+        outer_write_closure_([this]() { OnIocpWriteable(); }),
+        on_tcp_connect_locked_([this]() { OnTcpConnect(); }),
+        event_engine_(event_engine) {}
+
+  ~GrpcPolledFdWindows() override {
+    GRPC_ARES_RESOLVER_TRACE_LOG(
+        "fd:|%s| ~GrpcPolledFdWindows shutdown_called_: %d ", GetName(),
+        shutdown_called_);
+    grpc_core::CSliceUnref(read_buf_);
+    grpc_core::CSliceUnref(write_buf_);
+    GPR_ASSERT(read_closure_ == nullptr);
+    GPR_ASSERT(write_closure_ == nullptr);
+    if (!shutdown_called_) {
+      winsocket_->Shutdown(DEBUG_LOCATION, "~GrpcPolledFdWindows");
+    }
+  }
+
+  void RegisterForOnReadableLocked(
+      absl::AnyInvocable<void(absl::Status)> read_closure) override {
+    GPR_ASSERT(read_closure_ == nullptr);
+    read_closure_ = std::move(read_closure);
+    grpc_core::CSliceUnref(read_buf_);
+    GPR_ASSERT(!read_buf_has_data_);
+    read_buf_ = GRPC_SLICE_MALLOC(kReadBufferSize);
+    if (connect_done_) {
+      ContinueRegisterForOnReadableLocked();
+    } else {
+      GPR_ASSERT(pending_continue_register_for_on_readable_locked_ == false);
+      pending_continue_register_for_on_readable_locked_ = true;
+    }
+  }
+
+  void RegisterForOnWriteableLocked(
+      absl::AnyInvocable<void(absl::Status)> write_closure) override {
+    if (socket_type_ == SOCK_DGRAM) {
+      GRPC_ARES_RESOLVER_TRACE_LOG(
+          "fd:|%s| RegisterForOnWriteableLocked called", GetName());
+    } else {
+      GPR_ASSERT(socket_type_ == SOCK_STREAM);
+      GRPC_ARES_RESOLVER_TRACE_LOG(
+          "fd:|%s| RegisterForOnWriteableLocked called tcp_write_state_: %d "
+          "connect_done_: %d",
+          GetName(), tcp_write_state_, connect_done_);
+    }
+    GPR_ASSERT(write_closure_ == nullptr);
+    write_closure_ = std::move(write_closure);
+    if (!connect_done_) {
+      GPR_ASSERT(!pending_continue_register_for_on_writeable_locked_);
+      pending_continue_register_for_on_writeable_locked_ = true;
+      // Register an async OnTcpConnect callback here rather than when the
+      // connect was initiated, since we are now guaranteed to hold a ref of the
+      // c-ares wrapper before write_closure_ is called.
+      winsocket_->NotifyOnWrite(&on_tcp_connect_locked_);
+    } else {
+      ContinueRegisterForOnWriteableLocked();
+    }
+  }
+
+  bool IsFdStillReadableLocked() override { return read_buf_has_data_; }
+
+  void ShutdownLocked(absl::Status error) override {
+    GPR_ASSERT(!shutdown_called_);
+    if (absl::IsCancelled(error)) {
```

Comment:
Hmm, since we only call `Shutdown` during external cancellation, I think there might be a problem here:

Consider if:
- we have outstanding read and write callbacks pending
- writability fires so we write on the socket but hit an error
- c-ares loses interest in reading or writing on the socket
- we still have a readability callback pending

Because we'll call `Shutdown` immediately after seeing c-ares has no interest, we'll mark the fd node `already_shutdown = true`. But the readability callback will still be pending.

In this case, would the readability callback ever fire? Worried we would just hand indefinitely in this case.

Reply:
Yeah, good point. It seems like we need to slightly tweak the `ShutdownLocked()` API to return a bool indicating whether the shutdown is confirmed or not. I have made the change in https://github.com/grpc/grpc/commit/eb5641bbebabcf661c4e8341fd4eb06497902d6b. PTAL.

Reply:
I'm still wondering if we can find another way, I have some reservations about this approach:
- we're not looking at the return value of `ShutdownLocked`, which is a little hard to make sense of
- we need to carefully make sure that we invoke `ShutdownLocked` with a `Cancelled` status code if we want it to actually do anything *on windows*

Here is another idea:

What if we just got rid of the loop in [here](https://github.com/grpc/grpc/blob/a3f2e082024c3510635bb2e7584c3e4ebcaade7a/src/core/lib/event_engine/ares_resolver.cc#L418) ?

I.e., never remove fds from AresResolver fd node list, at least until we Orphan AresResolver.

In the iomgr based ares resolver, we need that loop because we don't invoke the application's callback until all fd read/write callbacks are flushed out. But in the EE ares resolver, I think we can just rely on AresResolver Orphan to shutdown all remaining fds anyways.

Then, we could remove the conditional logic we currently have around ShutdownLocked on windows. And I think posix would work fine.

Reply:
Are you suggesting something like this patch: https://gist.github.com/yijiem/1b8261f87ff1ec4a22d3410653d4373d? I tried the patch and multiple tests are failing.

This loop does something on Windows, we are shutting down the socket when there are no pending callbacks at: https://github.com/grpc/grpc/blob/a3f2e082024c3510635bb2e7584c3e4ebcaade7a/src/core/lib/event_engine/ares_resolver.cc#L429

It seems like without this logic, we are hitting assertions in https://github.com/grpc/grpc/blob/d77e5c0dd38f226a1e569cba309e82387a16f900/src/core/lib/event_engine/windows/win_socket.cc#L127 and https://github.com/grpc/grpc/blob/a3f2e082024c3510635bb2e7584c3e4ebcaade7a/src/core/lib/event_engine/ares_resolver.cc#L454.

Reply:
Ah, I was wrong about the loop shutting down the actual WinSocket (sorry my memory of this code is already blur). But it does update our book-keeping to not track that (which I think is making some difference here).